File: Documentation/4.1/Samples/java/src/test/java/net/ravendb/ClientApi/HowTo/SetupAggressiveCaching.java
package net.ravendb.ClientApi.HowTo;

import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.session.IDocumentSession;
import net.ravendb.client.documents.conventions.DocumentConventions;
import net.ravendb.client.primitives.CleanCloseable;

import java.time.Duration;
import java.util.List;

public class SetupAggressiveCaching {
    public SetupAggressiveCaching() {
        //region aggressive_cache_conventions
        try (IDocumentStore documentStore = new DocumentStore()) {
            DocumentConventions conventions = documentStore.getConventions();

            conventions.aggressiveCache().setDuration(Duration.ofMinutes(5));
            conventions.aggressiveCache().setMode(AggressiveCacheMode.TRACK_CHANGES);
            // Do your work here
        }
            //endregion
            
            //region aggressive_cache_global
            documentStore.aggressivelyCacheFor(Duration.ofMinutes(5));

            documentStore.aggressivelyCache(); // Defines the cache duration for 1 day
            //endregion

            try (IDocumentSession session = documentStore.openSession()) {
                //region aggressive_cache_load
                try (CleanCloseable cacheScope = session.advanced().getDocumentStore()
                        .aggressivelyCacheFor(Duration.ofMinutes(5))) {
                    Order user = session.load(Order.class, "orders/1");
                }
                //endregion

                //region disable_aggressive_cache
                try (CleanCloseable cacheScope = session.advanced().getDocumentStore()
                        .disableAggressiveCaching()) {
                    Order order = session.load(Order.class, "orders/1");
                }
                //endregion
            }

            try (IDocumentSession session = documentStore.openSession()) {
                //region aggressive_cache_query
                try (CleanCloseable cacheScope = session.advanced().getDocumentStore()
                        .aggressivelyCacheFor(Duration.ofMinutes(5))) {
                    List<Order> orders = session.query(Order.class)
                        .toList();
                }
                //endregion
                
                //region disable_change_tracking
                documentStore.aggressivelyCacheFor(Duration.ofMinutes(5), AggressiveCacheMode.DO_NOT_TRACK_CHANGES);

                // Disable change tracking for just one session:
                try (session.advanced().getDocumentStore().aggressivelyCacheFor(Duration.ofMinutes(5),
                    AggressiveCacheMode.DO_NOT_TRACK_CHANGES)) {
                }
                //endregion

                //region aggressive_cache_for_one_day_1
                try (CleanCloseable cacheScope = session
                    .advanced().getDocumentStore().aggressivelyCache()) {

                }
                //endregion

                //region aggressive_cache_for_one_day_2
                try (CleanCloseable cacheScope = session
                    .advanced().getDocumentStore().aggressivelyCacheFor(Duration.ofDays(1))) {

                }
                //endregion
            }
        }
    }

    private static class Order {
    }
}

File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/HowTo/SetupAggressiveCaching.cs
using System;
using System.Collections.Generic;
using System.Linq;

using Raven.Client.Documents;
using Raven.Client.Http;
using Raven.Documentation.Samples.Orders;
using Sparrow;

namespace Raven.Documentation.Samples.ClientApi.HowTo
{
    public class SetupAggressiveCaching
    {
        public SetupAggressiveCaching()
        {
            
            using (
            #region aggressive_cache_conventions
            var documentStore = new DocumentStore
            {
                Urls = new[] { "http://localhost:8080" },
                Database = "NorthWind",
                Conventions =
                {
                    AggressiveCache =
                    {
                        Duration = TimeSpan.FromMinutes(5),
                        Mode = AggressiveCacheMode.TrackChanges
                    }
                }
            }
            #endregion
            )
            {
                documentStore.Initialize();
                #region aggressive_cache_global

                documentStore.AggressivelyCacheFor(TimeSpan.FromMinutes(5));

                documentStore.AggressivelyCache(); // Defines the cache duration for 1 day

                #endregion
                using (var session = documentStore.OpenSession())
                {
                    

                    #region aggressive_cache_load
                    using (session.Advanced.DocumentStore.AggressivelyCacheFor(TimeSpan.FromMinutes(5)))
                    {
                        Order user = session.Load<Order>("orders/1");
                    }
                    #endregion

                    #region disable_aggressive_cache

                    using (session.Advanced.DocumentStore.DisableAggressiveCaching())
                    {
                        Order user = session.Load<Order>("orders/1");
                    }

                    #endregion
                }

                using (var session = documentStore.OpenSession())
                {
                    #region aggressive_cache_query
                    using (session.Advanced.DocumentStore.AggressivelyCacheFor(TimeSpan.FromMinutes(5)))
                    {
                        List<Order> users = session.Query<Order>().ToList();
                    }
                    #endregion

                    #region disable_change_tracking
                    documentStore.AggressivelyCacheFor(TimeSpan.FromMinutes(5), AggressiveCacheMode.DoNotTrackChanges);

                    //Disable change tracking for just one session:
                    using (session.Advanced.DocumentStore.AggressivelyCacheFor(TimeSpan.FromMinutes(5), 
                                                                               AggressiveCacheMode.DoNotTrackChanges))
                    { }
                    #endregion

                    #region aggressive_cache_for_one_day_1
                    using (session.Advanced.DocumentStore.AggressivelyCache())
                    { }
                    #endregion

                    #region aggressive_cache_for_one_day_2
                    using (session.Advanced.DocumentStore.AggressivelyCacheFor(TimeSpan.FromDays(1)))
                    { }
                    #endregion
                }

            }
        }
    }
}
